File: src/main/java/com/github/kory33/signvote/model/VotePoint.java
package com.github.kory33.signvote.model;

import com.github.kory33.signvote.constants.Formats;
import com.github.kory33.signvote.constants.VotePointDataFileKeys;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import lombok.Getter;
import lombok.Setter;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.block.BlockState;
import org.bukkit.block.Sign;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

/**
 * A class representing a sign of a vote point
 */
public class VotePoint {
    @Setter @Getter private String name;
    @Getter private Sign voteSign;

    public VotePoint(String name, Sign voteSign) {
        this.name = name;
        this.voteSign = voteSign;
    }

    public VotePoint(File votePointFIle) throws IllegalArgumentException, IOException {
        BufferedReader reader = Files.newBufferedReader(votePointFIle.toPath(), Formats.FILE_ENCODING);
        JsonObject jsonObject = (new JsonParser()).parse(reader).getAsJsonObject();
        reader.close();

        this.name = jsonObject.get(VotePointDataFileKeys.NAME).getAsString();

        String worldName = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_WORLD).getAsString();
        World world = Bukkit.getWorld(worldName);
        int signX = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_LOC_X).getAsInt();
        int signY = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_LOC_Y).getAsInt();
        int signZ = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_LOC_Z).getAsInt();

        if (world == null) {
            throw new IllegalArgumentException("Invalid file given! (Invalid world name " + worldName + ")");
        }

        BlockState state = world.getBlockAt(signX, signY, signZ).getState();
        if (!(state instanceof Sign)) {
            throw new IllegalStateException("Sign does not exist at the location in the saved data!");
        }

        this.voteSign = (Sign) state;
    }

    /**
     * Get the json-serialized data of this object.
     * @return a json file that contains
     */
    public JsonObject toJson() {
        JsonObject jsonObject = new JsonObject();

        jsonObject.addProperty(VotePointDataFileKeys.NAME, this.name);
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_WORLD, this.voteSign.getWorld().getName());
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_LOC_X, this.voteSign.getX());
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_LOC_Y, this.voteSign.getY());
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_LOC_Z, this.voteSign.getZ());

        return jsonObject;
    }
}

File: src/main/kotlin/com/github/kory33/signvote/vote/VotePoint.kt
package com.github.kory33.signvote.vote

import com.github.kory33.signvote.constants.VotePointDataFileKeys
import com.github.kory33.signvote.utils.FileUtils
import com.google.gson.JsonObject
import org.bukkit.Bukkit
import org.bukkit.block.Sign
import java.io.File
import java.io.IOException

/**
 * A class representing a sign of a vote point
 */
class VotePoint {
    var name: String
    var voteSign: Sign
        private set

    constructor(name: String, voteSign: Sign) {
        this.name = name
        this.voteSign = voteSign
    }

    @Throws(IllegalArgumentException::class, IOException::class)
    constructor(votePointFIle: File) {
        val jsonObject = FileUtils.readJSON(votePointFIle)

        this.name = jsonObject.get(VotePointDataFileKeys.NAME).asString

        val worldName = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_WORLD).asString
        val world = Bukkit.getWorld(worldName)
        val signX = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_LOC_X).asInt
        val signY = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_LOC_Y).asInt
        val signZ = jsonObject.get(VotePointDataFileKeys.VOTE_SIGN_LOC_Z).asInt

        if (world == null) {
            throw IllegalArgumentException("Invalid file given! (Invalid world name $worldName)")
        }

        val state = world.getBlockAt(signX, signY, signZ).state as? Sign ?: throw IllegalStateException("Sign does not exist at the location in the saved data!")

        this.voteSign = state
    }

    /**
     * Get the json-serialized data of this object.
     * @return a json file that contains
     */
    fun toJson(): JsonObject {
        val jsonObject = JsonObject()

        jsonObject.addProperty(VotePointDataFileKeys.NAME, this.name)
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_WORLD, this.voteSign.world.name)
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_LOC_X, this.voteSign.x)
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_LOC_Y, this.voteSign.y)
        jsonObject.addProperty(VotePointDataFileKeys.VOTE_SIGN_LOC_Z, this.voteSign.z)

        return jsonObject
    }
}
